### PR TITLE
Elastic Integration

### PR DESCRIPTION
**Pull request by mbliakharskyi**
## Description

Please include a summary of the changes and issues addressed in this PR. Also, explain the motivation and context of each update. List any dependencies or prerequisites needed for implementing this change.

## Checklist:

Before you create this PR, confirm all the requirements listed below by checking the checkboxes like this (`[x]`).

- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests to cover the new feature or fixes.

### CODE REVIEW

File: apps/elastic/src/common/apiclient.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { ElasticError } from '@/connectors/commons/error';
+
+export const getElasticApiClient = () => {
```
Comment (elbalexandre) [inline apps/elastic/src/common/apiclient.ts:3]: Please don't do that kind of factorisation. It act as a black box when reading the connector code.

We might give tools to write less code later, but we need to discuss the implementation first.

File: apps/elastic/src/connectors/users.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    throw new ElasticError('Could not retrieve account id');
+  }
+  const { id: accountId } = accounts[0];
+
+  return {
+    accountId: accountId.toString(),
+  };
+};
+
+export const deleteUsers = async ({ userIds, accountId, apiKey }: DeleteUsersParams) => {
```
Comment (elbalexandre) [inline apps/elastic/src/connectors/users.ts:103]: Let's delete user one by one (even if the API allows multiple user deletion at once)
```suggestion
export const deleteUser = async ({ userId, accountId, apiKey }: DeleteUsersParams) => {
```

File: apps/elastic/src/connectors/users.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    method: 'DELETE',
+    headers: {
+      'Content-Type': 'application/json',
+      Authorization: `ApiKey ${apiKey}`,
+    },
+  });
+
+  if (!response.ok && response.status !== 400) {
+    throw new ElasticError(`Could not delete user with Id: ${userIds}`, { response });
+  }
```
Comment (elbalexandre) [inline apps/elastic/src/connectors/users.ts:116]:
```suggestion
const elasticErrorDataSchema = z.object({
  errors: z.array(z.object({
    code: z.string().nullish(),
    message: z.string().nullish()
  }))
})
// ...
  if (response.status === 400) {
    const errorDataResult = elasticErrorDataSchema.safeParse(await response.json())
    if (errorDataResult.success && errorDataResult.data.errors.at(0)?.code === 'organization.membership_not_found') {
      return;
    }
  }

  if (!response.ok) {
    throw new ElasticError(`Could not delete user with Id: ${userId}`, { response });
  }
```

File: apps/elastic/src/app/api/webhook/elba/users/delete/service.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { inngest } from '@/inngest/client';
+
+export const deleteUsers = async ({
```
Comment (elbalexandre) [inline apps/elastic/src/app/api/webhook/elba/users/delete/service.ts:3]: It should be `userId` here and in the inngest event (so in the function deleteUser too)

Comment (mbliakharskyi) [reply apps/elastic/src/app/api/webhook/elba/users/delete/service.ts:3]: okay.

File: apps/elastic/src/inngest/functions/users/delete-users.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    const [organisation] = await db
+      .select({
+        apiKey: Organisation.apiKey,
+        accountId: Organisation.accountId,
+      })
+      .from(Organisation)
+      .where(eq(Organisation.id, organisationId));
+
+    if (!organisation) {
+      throw new NonRetriableError(`Could not retrieve ${userId}`);
```
Comment (elbalexandre) [inline apps/elastic/src/inngest/functions/users/delete-users.ts:31]:
```suggestion
      throw new NonRetriableError(`Could not retrieve organisation with id=${organisationId}`);
```

**Review by elbalexandre — APPROVED** [review]

**Pull request merged**